lib/internal/bootstrap/loaders.js

```diff
@@ -216,12 +217,19 @@ class NativeModule {
   }
 
   // Used by user-land module loaders to compile and load builtins.
-  compileForPublicLoader() {
+  compileForPublicLoader(isPreloading) {
     if (!this.canBeRequiredByUsers) {
       // No code because this is an assertion against bugs
       // eslint-disable-next-line no-restricted-syntax
       throw new Error(`Should not compile ${this.id} for public use`);
     }
+    if (isPreloading) {
+      ObjectDefineProperty(
+        this, 'isPreloading',
+        ReflectGetOwnPropertyDescriptor(
+          nativeModuleRequire('internal/modules/cjs/loader').Module.prototype,
+          'isPreloading'));
+    }
```

Comment:
Maybe instead of defining this here, this property should be added to the `NativeModule.prototype` alongside: <https://github.com/nodejs/node/blob/b74793fce723d50417bf9256373ecc3f0640d807/lib/internal/modules/cjs/loader.js#L236-L238>

---

The best way to do this would probably be to extract the getter definition:
```js
const isPreloadingGetter = ObjectGetOwnPropertyDescriptor({
  get isPreloading() { return isPreloading; },
}, 'isPreloading').get;

ObjectDefineProperty(Module.prototype, 'isPreloading', { get: isPreloadingGetter });
ObjectDefineProperty(NativeModule.prototype, 'isPreloading', { get: isPreloadingGetter });
```

---

This is already being done for `Event.<instance>.isTrusted`: <https://github.com/nodejs/node/blob/f658bd1e8f7cd0a328c10c8d2a4eea6cf113122b/lib/internal/event_target.js#L77-L82><https://github.com/nodejs/node/blob/f658bd1e8f7cd0a328c10c8d2a4eea6cf113122b/lib/internal/event_target.js#L102-L107>

Reply:
Good catch, since `NativeModule` is required by `/internal/modules/cjs/loader`, it simplifies the logic. What do you think of reusing the same property descriptor object:

```js
const isPreloadingDescriptor = {
  get() { return isPreloading; },
};

ObjectDefineProperty(Module.prototype, 'isPreloading', isPreloadingGetter);
ObjectDefineProperty(NativeModule.prototype, 'isPreloading', isPreloadingGetter);
```

Wouldn't it be more efficient to parse and compute?

Reply:
Mine has the getter have a `name` of `"get isPreloading"`, whereas yours has a `name` of just `"get"`.

Reply:
Right, but considering `get` was already the name of the getter, I'd rather keep it as is.